L1Trigger/Phase2L1ParticleFlow/plugins/L1TPFProducer.cc

```diff
@@ -0,0 +1,381 @@
+      l1regions_(iConfig),
+      l1pfalgo_(nullptr),
+      l1pualgo_(nullptr),
+      fRegionDump(nullptr),
+      fRegionCOE(nullptr),
+      debugEta_(iConfig.getUntrackedParameter<double>("debugEta", 0)),
+      debugPhi_(iConfig.getUntrackedParameter<double>("debugPhi", 0)),
+      debugR_(iConfig.getUntrackedParameter<double>("debugR", -1)) {
+  produces<l1t::PFCandidateCollection>("PF");
+  produces<l1t::PFCandidateCollection>("Puppi");
+
+  //produces<l1t::PFCandidateCollection>("RawEmCalo");
+  //produces<l1t::PFCandidateCollection>("RawCalo");
+
+  produces<l1t::PFCandidateCollection>("EmCalo");
+  produces<l1t::PFCandidateCollection>("Calo");
+  produces<l1t::PFCandidateCollection>("TK");
+  produces<l1t::PFCandidateCollection>("TKVtx");
+
+  produces<float>("z0");
+
+  for (auto& tag : iConfig.getParameter<std::vector<edm::InputTag>>("emClusters")) {
+    emCands_.push_back(consumes<l1t::PFClusterCollection>(tag));
+  }
+  for (auto& tag : iConfig.getParameter<std::vector<edm::InputTag>>("hadClusters")) {
+    hadCands_.push_back(consumes<l1t::PFClusterCollection>(tag));
+  }
+
+  const std::string& algo = iConfig.getParameter<std::string>("pfAlgo");
+  if (algo == "PFAlgo3") {
+    l1pfalgo_.reset(new l1tpf_impl::PFAlgo3(iConfig));
+  } else if (algo == "PFAlgo2HGC") {
+    l1pfalgo_.reset(new l1tpf_impl::PFAlgo2HGC(iConfig));
+  } else if (algo == "BitwisePFAlgo") {
+    l1pfalgo_.reset(new l1tpf_impl::BitwisePFAlgo(iConfig));
+  } else
+    throw cms::Exception("Configuration", "Unsupported PFAlgo");
+
+  const std::string& pualgo = iConfig.getParameter<std::string>("puAlgo");
+  if (pualgo == "Puppi") {
+    l1pualgo_.reset(new l1tpf_impl::PuppiAlgo(iConfig));
+  } else if (pualgo == "LinearizedPuppi") {
+    l1pualgo_.reset(new l1tpf_impl::LinearizedPuppiAlgo(iConfig));
+  } else
+    throw cms::Exception("Configuration", "Unsupported PUAlgo");
+
+  std::string vtxAlgo = iConfig.getParameter<std::string>("vtxAlgo");
+  if (vtxAlgo == "TP")
+    vtxAlgo_ = l1tpf_impl::PUAlgoBase::TPVtxAlgo;
+  else if (vtxAlgo == "old")
+    vtxAlgo_ = l1tpf_impl::PUAlgoBase::OldVtxAlgo;
+  else if (vtxAlgo == "external") {
+    vtxAlgo_ = l1tpf_impl::PUAlgoBase::ExternalVtxAlgo;
+    const std::string& vtxFormat = iConfig.getParameter<std::string>("vtxFormat");
+    if (vtxFormat == "TkPrimaryVertex") {
+      extTkVtx_ = consumes<std::vector<l1t::TkPrimaryVertex>>(iConfig.getParameter<edm::InputTag>("vtxCollection"));
+    } else
+      throw cms::Exception("Configuration") << "Unsupported vtxFormat " << vtxFormat << "\n";
+  } else
+    throw cms::Exception("Configuration") << "Unsupported vtxAlgo " << vtxAlgo << "\n";
+
+  for (const std::string& label : l1pualgo_->puGlobalNames()) {
+    produces<float>(label);
+  }
+
+  std::string dumpFileName = iConfig.getUntrackedParameter<std::string>("dumpFileName", "");
+  if (!dumpFileName.empty()) {
+    fRegionDump = fopen(dumpFileName.c_str(), "wb");
```

Comment:
production of dump files is a rare occurrence and by experienced users, so I wouldn't complicate the producer for that.
(but if it's simple I can put a check that CMSSW is being run single-threaded and throw an exception otherwise)

Reply:
@makortel please advise about the thread safety concern

Reply:
As far as I can tell, `fopen()`, `fwrite()`, `fclose()` are thread safe (at least) when operating on different files. I would be fine with a check if this piece will not be run in production. I think the check should be on the number of EDM streams though (multiple threads are fine as long as there is only a single instance of this class), assuming this class does not do any kind of internal parallelization (with TBB).

Reply:
Maybe the simplest thing is just to require a mutex when ever `fRegionDump` is begin called?

Reply:
@makortel is there a simple way to check the number of streams?

Reply:
One option would be to move the file opening to `beginStream()`, and throw an exception if the argument `StreamID != 0`. Alternative to the exception could be to add the `StreamID` to the file name and have a separate file for each EDM stream.